llvm/utils/TableGen/DAGISelMatcherGen.cpp

```diff
@@ -602,6 +601,7 @@ bool MatcherGen::EmitMatcherCode(unsigned Variant) {
 
     // Emit a CheckComplexPat operation, which does the match (aborting if it
     // fails) and pushes the matched operands onto the recorded nodes list.
+    CGP.increaseComplexPatternUsage(CP);
```

Comment:
Could you have done this before the actual emission to keep the const during emission?

Reply:
Thanks!
I have changed the implementation, now we don't need to record the usage when creating `XXXMatcher` nodes. Instead, we will do statistic just before emission (with a single pass iteration over the whole MatcherTable). I do this for two reasons:
* The previous implementation is too intrusive, and, as you see, we lost the const qualifier.
* There are some optimizations after creating `XXXMatcher` nodes and some nodes may be deleted, so we should do the statistic after optimizations.

The implementation is much easier to review now.